File: scripts/eval/compare.py
"""Compare two eval results files: aggregate deltas + per-query regression list.

    uv run python -m scripts.eval.compare scripts/eval/results/baseline-*.json \
                                                scripts/eval/results/phase1-*.json

Matches queries by `id`; reports hit@1/3/8 and MRR deltas and lists per-query
regressions (rank got worse or a hit was lost) plus improvements. Exit 0
either way — it is a measurement tool, not a gate — but --fail-on-regression
gives CI a gate later.
"""

from __future__ import annotations

import argparse
import json
import sys
from pathlib import Path


def load_rows(path: str) -> dict[str, dict]:
    """{id -> row} from a run.py results JSON; errors/empty rows included."""
    data = json.loads(Path(path).read_text(encoding="utf-8"))
    return {row["id"]: row for row in data["rows"]}


def hit_rank(row: dict) -> int | None:
    """Effective hit rank: None for error/miss rows; misses rank worse than
    any real rank (top_k + 1) so deltas and regressions stay well-defined."""
    if row.get("error") or not row.get("results"):
        return None
    return row.get("hit_rank")


def classify(base_row: dict, cand_row: dict, top_k: int) -> str:
    """Compare one query pair: 'regression', 'improvement', or 'unchanged'."""
    base_rank, cand_rank = hit_rank(base_row), hit_rank(cand_row)
    if base_rank == cand_rank:
        return "unchanged"
    base_key = base_rank if base_rank is not None else top_k + 1
    cand_key = cand_rank if cand_rank is not None else top_k + 1
    return "regression" if cand_key > base_key else "improvement"


def compare(base_path: str, cand_path: str, top_k: int) -> dict:
    """Deltas + per-query regression/improvement lists, matched by id."""
    base, cand = json.loads(Path(base_path).read_text(encoding="utf-8")), json.loads(
        Path(cand_path).read_text(encoding="utf-8")
    )
    base_rows, cand_rows = load_rows(base_path), load_rows(cand_path)
    regressions, improvements = [], []
    shared = sorted(set(base_rows) & set(cand_rows))
    for query_id in shared:
        verdict = classify(base_rows[query_id], cand_rows[query_id], top_k)
        if verdict == "regression":
            regressions.append(query_id)
        elif verdict == "improvement":
            improvements.append(query_id)
    base_summary, cand_summary = base.get("summary", {}), cand.get("summary", {})
    deltas = {}
    for metric in ("hit_at_1", "hit_at_3", "hit_at_8", "mrr"):
        if metric in base_summary and metric in cand_summary:
            deltas[metric] = round(cand_summary[metric] - base_summary[metric], 4)
    return {
        "matched": len(shared),
        "only_in_base": sorted(set(base_rows) - set(cand_rows)),
        "only_in_candidate": sorted(set(cand_rows) - set(base_rows)),
        "deltas": deltas,
        "regressions": regressions,
        "improvements": improvements,
    }


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(
        description="Compare two eval results files (A vs B)."
    )
    parser.add_argument("base", help="Baseline results JSON")
    parser.add_argument("candidate", help="Candidate results JSON")
    parser.add_argument(
        "--fail-on-regression",
        action="store_true",
        help="Exit non-zero when any query regressed (for CI gating later)",
    )
    args = parser.parse_args(argv)
    top_k = json.loads(Path(args.base).read_text(encoding="utf-8")).get("top_k", 8)
    report = compare(args.base, args.candidate, top_k)

    print(f"matched queries: {report['matched']}")
    if report["only_in_base"]:
        print(f"only in base: {', '.join(report['only_in_base'])}")
    if report["only_in_candidate"]:
        print(f"only in candidate: {', '.join(report['only_in_candidate'])}")
    for metric, delta in report["deltas"].items():
        print(f"{metric}: {delta:+.4f}")
    for title, ids in (("regressions", report["regressions"]),
                       ("improvements", report["improvements"])):
        print(f"{title}: {len(ids)}")
        for query_id in ids:
            print(f"  - {query_id}")
    if args.fail_on_regression and report["regressions"]:
        print(f"error: {len(report['regressions'])} regression(s)", file=sys.stderr)
        return 1
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: cmd/lybrix-eval/compare.go
package main

import (
	"encoding/json"
	"flag"
	"fmt"
	"os"
	"sort"
	"strconv"
)

// Compare two eval results files: aggregate deltas + per-query regression
// list. Port of scripts/eval/compare.py. Exit 0 either way — it is a
// measurement tool, not a gate — but --fail-on-regression gives CI a gate.

// loadRows reads {id -> row} from a run results JSON; errors/empty rows
// included.
func loadRows(path string) (map[string]map[string]any, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	var doc struct {
		Rows []map[string]any `json:"rows"`
	}
	if err := json.Unmarshal(data, &doc); err != nil {
		return nil, err
	}
	out := map[string]map[string]any{}
	for _, row := range doc.Rows {
		out[toStr(row["id"])] = row
	}
	return out, nil
}

// HitRank is the effective hit rank: nil for error/miss rows; misses rank
// worse than any real rank (top_k + 1) so deltas and regressions stay
// well-defined.
func HitRank(row map[string]any) *int {
	if _, isErr := row["error"]; isErr {
		return nil
	}
	rs, ok := row["results"].([]map[string]any)
	if !ok || len(rs) == 0 {
		return nil
	}
	if rank, ok := row["hit_rank"].(*int); ok {
		return rank
	}
	return nil
}

// Classify compares one query pair: "regression", "improvement", or
// "unchanged".
func Classify(baseRow, candRow map[string]any, topK int) string {
	baseRank, candRank := HitRank(baseRow), HitRank(candRow)
	if sameRank(baseRank, candRank) {
		return "unchanged"
	}
	baseKey := topK + 1
	if baseRank != nil {
		baseKey = *baseRank
	}
	candKey := topK + 1
	if candRank != nil {
		candKey = *candRank
	}
	if candKey > baseKey {
		return "regression"
	}
	return "improvement"
}

func sameRank(a, b *int) bool {
	if a == nil && b == nil {
		return true
	}
	if a == nil || b == nil {
		return false
	}
	return *a == *b
}

// summaryValue reads a metric from a summary with a hit_at_top_k
// back-compat fallback (old result files carry hit_at_8 under
// f"hit_at_{top_k}").
func summaryValue(summary map[string]any, metric string) (float64, bool) {
	if v, ok := summary[metric].(float64); ok {
		return v, true
	}
	if metric == "hit_at_top_k" {
		if v, ok := summary["hit_at_8"].(float64); ok {
			return v, true
		}
	}
	return 0, false
}

func compareCmd(argv []string) error {
	fs := flag.NewFlagSet("compare", flag.ExitOnError)
	failOnRegression := fs.Bool("fail-on-regression", false, "exit non-zero when any query regressed")
	if err := fs.Parse(argv); err != nil {
		return err
	}
	if fs.NArg() != 2 {
		return fmt.Errorf("usage: lybrix-eval compare <base.json> <candidate.json>")
	}
	basePath, candPath := fs.Arg(0), fs.Arg(1)

	baseDoc, err := readDoc(basePath)
	if err != nil {
		return err
	}
	topK := 8
	if baseDoc.TopK != 0 {
		topK = baseDoc.TopK
	}
	baseRows, err := loadRows(basePath)
	if err != nil {
		return err
	}
	candRows, err := loadRows(candPath)
	if err != nil {
		return err
	}

	var regressions, improvements []string
	shared := []string{}
	baseOnly := []string{}
	candOnly := []string{}
	for id := range baseRows {
		if _, ok := candRows[id]; !ok {
			baseOnly = append(baseOnly, id)
		}
	}
	for id := range candRows {
		if _, ok := baseRows[id]; !ok {
			candOnly = append(candOnly, id)
		} else {
			shared = append(shared, id)
		}
	}
	sort.Strings(shared)
	for _, id := range shared {
		switch Classify(baseRows[id], candRows[id], topK) {
		case "regression":
			regressions = append(regressions, id)
		case "improvement":
			improvements = append(improvements, id)
		}
	}

	fmt.Printf("matched queries: %d\n", len(shared))
	if len(baseOnly) > 0 {
		fmt.Printf("only in base: %s\n", joinComma(baseOnly))
	}
	if len(candOnly) > 0 {
		fmt.Printf("only in candidate: %s\n", joinComma(candOnly))
	}
	for _, metric := range []string{"hit_at_1", "hit_at_3", "hit_at_top_k", "hit_at_8", "mrr"} {
		b, okB := summaryValue(baseDoc.Summary, metric)
		c, okC := summaryValue(readCandidateSummary(candPath), metric)
		if okB && okC {
			delta := c - b
			fmt.Printf("%s: %s\n", metric, fmtDelta(delta))
		}
	}
	for _, pair := range []struct {
		title string
		ids   []string
	}{{"regressions", regressions}, {"improvements", improvements}} {
		fmt.Printf("%s: %d\n", pair.title, len(pair.ids))
		for _, id := range pair.ids {
			fmt.Printf("  - %s\n", id)
		}
	}
	if *failOnRegression && len(regressions) > 0 {
		fmt.Fprintf(os.Stderr, "error: %d regression(s)\n", len(regressions))
		os.Exit(1)
	}
	return nil
}

// resultsDoc mirrors the run results JSON top-level shape.
type resultsDoc struct {
	TopK    int            `json:"top_k"`
	Summary map[string]any `json:"summary"`
}

func readDoc(path string) (*resultsDoc, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	var doc resultsDoc
	if err := json.Unmarshal(data, &doc); err != nil {
		return nil, err
	}
	return &doc, nil
}

func readCandidateSummary(path string) map[string]any {
	doc, err := readDoc(path)
	if err != nil {
		return nil
	}
	return doc.Summary
}

func joinComma(xs []string) string {
	out := ""
	for i, x := range xs {
		if i > 0 {
			out += ", "
		}
		out += x
	}
	return out
}

func fmtDelta(d float64) string {
	sign := "+"
	if d < 0 {
		sign = "-"
		d = -d
	}
	return sign + strconv.FormatFloat(d, 'f', 4, 64)
}
